geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/cache/ThreadLocalCache.scala

```diff
@@ -0,0 +1,146 @@
+
+  private val caches = new ThreadLocal[Cache[K, V]]() {
+    override def initialValue(): Cache[K, V] = {
+      val cache = Caffeine.newBuilder().expireAfterAccess(expiry.toMillis, TimeUnit.MILLISECONDS).build[K, V]()
+      // this will always succeed as our queue is unbounded
+      references.offer((Thread.currentThread().getId, new WeakReference(cache)))
+      cache
+    }
+  }
+
+  private val cleanup = executor.scheduleWithFixedDelay(this, expiry.toMillis, expiry.toMillis, TimeUnit.MILLISECONDS)
+
+  // show the class name for toString
+  override def stringPrefix : String = "ThreadLocalCache"
+
+  override def get(key: K): Option[V] = Option(caches.get.getIfPresent(key))
+
+  override def getOrElseUpdate(key: K, op: => V): V = {
+    val cached = caches.get.getIfPresent(key)
+    if (cached != null) { cached } else {
+      val value = op
+      caches.get.put(key, value)
+      value
+    }
+  }
+
+  override def +=(kv: (K, V)): this.type = {
+    caches.get.put(kv._1, kv._2)
+    this
+  }
+
+  override def -=(key: K): this.type = {
+    caches.get.invalidate(key)
+    this
+  }
+
+  override def empty: ThreadLocalCache[K, V] = new ThreadLocalCache[K, V](expiry)
+
+  override def iterator: Iterator[(K, V)] = caches.get.asMap.asScala.iterator
+
+  /**
+   * Gets an iterator across all thread-local values, not just the current thread
+   *
+   * @return iterator of (thread-id, key, value)
+   */
+  def globalIterator: Iterator[(Long, K, V)] = {
+    references.iterator.asScala.flatMap { case (id, ref) =>
+      val cache = ref.get
+      if (cache == null) { Iterator.empty } else {
+        cache.asMap().asScala.iterator.map { case (k, v) => (id, k, v) }
+      }
+    }
+  }
+
+  /**
+   * Gets the estimated total size across all thread-local values
+   *
+   * @return
+   */
+  def estimatedGlobalSize: Long = {
+    var size = 0L
+    references.iterator.asScala.foreach { case (_, ref) =>
+      val cache = ref.get
+      if (cache != null) {
+        size += cache.estimatedSize()
+      }
+    }
+    size
+  }
+
+  override def run(): Unit = {
+    val iter = references.iterator()
+    while (iter.hasNext) {
+      val cache = iter.next._2.get
+      if (cache == null) {
+        // cache has been GC'd, remove our reference to it
+        iter.remove()
+      } else {
+        cache.cleanUp()
+      }
+    }
+  }
+
+  override def close(): Unit = {
+    cleanup.cancel(true)
+    val iter = references.iterator()
+    while (iter.hasNext) {
+      val cache = iter.next._2.get
+      if (cache != null) {
+        cache.asMap().clear()
+      }
+      iter.remove()
+    }
+  }
+}
+
+object ThreadLocalCache {
+  // use a 2 thread executor service for all the caches - we only use a handful across the code base
+  private val executor = MoreExecutors.getExitingScheduledExecutorService {
+    Executors.newScheduledThreadPool(2).asInstanceOf[ScheduledThreadPoolExecutor]
```

Comment:
Any reason to use 2 instead of 1?

Reply:
1 seemed too few, 3 seemed too many?

Reply:
I can be onboard with 2 being greater than 1, but not much greater.